zetta_utils/layer/volumetric/tensorstore/backend.py

```diff
@@ -0,0 +1,307 @@
+                "Data written to CloudVolume backend must be in `cxyz` dimension format, "
+                f"but got a tensor of with ndim == {data_np.ndim}"
+            )
+
+        ts = _get_ts_at_resolution(self.path, self.cache_bytes_limit, str(list(idx.resolution)))
+        slices = idx.to_slices()
+        ts[slices] = data_final
+
+    def with_changes(self, **kwargs) -> TSBackend:
+        """Currently untyped. Supports:
+        "name" = value: str
+        "allow_cache" = value: Union[bool, str] - must be False for TensorStoreBackend, ignored
+        "enforce_chunk_aligned_writes" = value: bool - must be False for TensorStoreBackend
+        "voxel_offset_res" = (voxel_offset, resolution): Tuple[Vec3D[int], Vec3D]
+        "chunk_size_res" = (chunk_size, resolution): Tuple[Vec3D[int], Vec3D]
+        """
+        # TODO: implement proper allow_cache logic
+        assert self.info_spec is not None
+
+        info_spec = deepcopy(self.info_spec)
+
+        implemented_keys = [
+            "name",
+            "allow_cache",
+            "enforce_chunk_aligned_writes",
+            "voxel_offset_res",
+            "chunk_size_res",
+        ]
+        keys_to_kwargs = {"name": "path"}
+        keys_to_infospec_fn = {
+            "voxel_offset_res": info_spec.set_voxel_offset,
+            "chunk_size_res": info_spec.set_chunk_size,
+        }
+        keys_to_assert = {"enforce_chunk_aligned_writes": False}
+        evolve_kwargs = {}
+        for k, v in kwargs.items():
+            if k not in implemented_keys:
+                raise KeyError(f"key `{k}` received, expected one of `{implemented_keys}`")
+            if k in keys_to_kwargs:
+                evolve_kwargs[keys_to_kwargs[k]] = v
+            if k in keys_to_infospec_fn:
+                keys_to_infospec_fn[k](v)
+            if k in keys_to_assert:
+                if v != keys_to_assert[k]:
+                    raise ValueError(
+                        f"key `{k}` received with value `{v}`, but is required to be "
+                        f"`{keys_to_assert[k]}`"
+                    )
+        # must clear the TS cache since the TS cache is separate from the info cache
+        _clear_ts_cache(self.path)
+        if "name" in kwargs:
+            _clear_ts_cache(kwargs["name"])
+
+        return attrs.evolve(
+            self,
+            **evolve_kwargs,
+            info_spec=info_spec,
+            on_info_exists="overwrite",
+        )
+
+    def get_voxel_offset(self, resolution: Vec3D) -> Vec3D[int]:
+        ts = _get_ts_at_resolution(self.path, self.cache_bytes_limit, str(list(resolution)))
+        return Vec3D[int](*ts.chunk_layout.grid_origin[0:3])
+
+    def get_chunk_size(self, resolution: Vec3D) -> Vec3D[int]:
+        ts = _get_ts_at_resolution(self.path, self.cache_bytes_limit, str(list(resolution)))
+        return Vec3D[int](*ts.chunk_layout.read_chunk.shape[0:3])
+
+    def get_chunk_aligned_index(  # pragma: no cover
```

Comment:
This should be tested

Reply:
I guess it's not just delegating to cloudvolume as in the cloudvol backend.

Reply:
Wait, isn't this just delegation to `idx.bbox.snapped`? That's pretty well tested.

Reply:
As is, it's quite a bit of logic.

If it was refactored to have only 1 if that throws exception `TensorStore backend only supports  'expand', 'shrink'`, and then delegates to `bbox.snapped` unconditionally, then it would be OK not to test.

Reply:
Two `if`'s for exceptions is also fine